### PR TITLE
Waypoint updater simple

### PR DESCRIPTION
**Pull request by lorenzolucido**
[#1] This first version of waypoint_updater simply takes in the static waypoints we want to follow, subscribe to the current position of the car and output the next 200 waypoints relatively to the current car position.

### CODE REVIEW

**Comment by agotterba** [comment]
I tried driving around the simulation track with this, and ran into some issues.  The first thing that happened was that sometimes, the string of waypoints would jump to be ahead of the car:

waypoints_ahead

On my first run, eventually, the waypoints stopped updating; I could drive circles around the string of waypoints without it changing:

waypoint_stop

On the second and third runs, the waypoints disappeared.  I piped the /current_pose and /final_waypoints topics to the same log file; they aren't labelled, but towards the end of the log (when I lost the waypoints), the current_pose posts have ' seq: 8' thousand something, while the final_waypoints posts are at 'seq: 7' hundred something (current_pose also has frame_id: "/world" while final_waypoints has frame_id: '', but final_waypoints lists frame_id for every waypoint, so if you want to search for that, specify 2 spaces after the beginning of the line).  seq: 754 of final_waypoints looks fine, but 755 is empty:

---
header: 
  seq: 755
  stamp: 
    secs: 0
    nsecs:         0
  frame_id: ''
waypoints: []
---

I've uploaded the full log (zipped; it expands to 72M):

[pose_waypoints.zip](https://github.com/clockworkbits/CarND-Capstone/files/1800083/pose_waypoints.zip)

These problems don't happen at exactly the same place; I wonder if it has something to do with crossing a 180 degree heading.  I did find that when making a wide turn at the start of the simulation (but less than 90 degrees), the waypoints disappeared.

File: ros/src/waypoint_updater/waypoint_updater.py [hunk 1]
```diff
@@ -62,16 +72,51 @@ def get_waypoint_velocity(self, waypoint):
...
+
+    def update_next_waypoint(self):
+        it = 0
+        nb_waypoints = len(self.static_waypoints.waypoints)
+        while not self.waypoint_is_ahead(self.static_waypoints.waypoints[self.next_waypoint % nb_waypoints]) and \
+                it < UPDATE_MAX_ITER:
+            self.next_waypoint += 1  # We look at the next one
+
+        # Searching the next waypoint in the full path takes much longer, we want to avoid it as much as possible
+        if it == UPDATE_MAX_ITER:
```
Comment (clockworkbits) [inline ros/src/waypoint_updater/waypoint_updater.py]: It seems that the `it` variable is never incremented, so `search_next_waypoint()` is never called.

File: ros/src/waypoint_updater/waypoint_updater.py [hunk 2]
```diff
@@ -62,16 +72,51 @@ def get_waypoint_velocity(self, waypoint):
...
+            self.search_next_waypoint()
+
+    def search_next_waypoint(self):
+        self.next_waypoint = 0
+        # TODO: Rewrite this in a clearer way
```
Comment (clockworkbits) [inline ros/src/waypoint_updater/waypoint_updater.py]: Pull Requests should avoid `TODOs`. A good practice is either do it now or leave it, but create an issue in GitHub and have a reference to it here as well. So the `TODO` is not forgotten.

Comment (clockworkbits) [reply ros/src/waypoint_updater/waypoint_updater.py]: You can think of @agotterba suggestion to use https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.spatial.KDTree.html . This could be a separate task.

File: ros/src/waypoint_updater/waypoint_updater.py [hunk 3]
```diff
@@ -62,16 +72,51 @@ def get_waypoint_velocity(self, waypoint):
...
+        # We basically search among all static waypoints the closest waypoint ahead
+        for i in range(len(self.static_waypoints)):
+            if self.waypoint_is_ahead(self.static_waypoints.waypoints[i]) and \
+                    self.distance_to_previous(self.static_waypoints.waypoints[i].pose) < \
+                    self.distance_to_previous(self.static_waypoints.waypoints[self.next_waypoint].pose):
+                self.next_waypoint = i
```
Comment (clockworkbits) [inline ros/src/waypoint_updater/waypoint_updater.py]: probably you can brake the loop when you find what the next waypoint.

File: ros/src/waypoint_updater/waypoint_updater.py (hunk 2 above)
Comment (agotterba) [reply ros/src/waypoint_updater/waypoint_updater.py]: Regarding scipy.spatial.KDTree, One issue that waypoint_updater has is that the point it's looking for should be in front of the car; tl_detector only has to look for the nearest point.  KDTree.query lets you ask for the nearest several points by setting the k argument greater than 1.  You can ask for a few points (4? 12?) to make sure at least one is in front of the car, and iterate through them until you find one that's in front (the points are returned sorted from nearest to farthest).

File: ros/src/waypoint_updater/waypoint_updater.py (hunk 1 above)
Comment (lorenzolucido) [reply ros/src/waypoint_updater/waypoint_updater.py]: Thanks, fixed that locally

File: ros/src/waypoint_updater/waypoint_updater.py (hunk 2 above)
Comment (lorenzolucido) [reply ros/src/waypoint_updater/waypoint_updater.py]: Exactly, this is why I haven't used KDTree eventually.

File: ros/src/waypoint_updater/waypoint_updater.py (hunk 3 above)
Comment (lorenzolucido) [reply ros/src/waypoint_updater/waypoint_updater.py]: But how do I know that this is the closest waypoint ahead if I stop early ?

File: ros/src/CMakeLists.txt [hunk 4]
```diff
@@ -1 +0,0 @@
-/opt/ros/kinetic/share/catkin/cmake/toplevel.cmake
```
Comment (clockworkbits) [inline ros/src/CMakeLists.txt]: Is this change intentional?

File: ros/src/waypoint_updater/waypoint_updater.py [hunk 5]
```diff
@@ -62,16 +81,62 @@ def get_waypoint_velocity(self, waypoint):
...
+        angle = normalize_angle(yaw-heading)
+        return True if math.fabs(angle) < math.pi/4. else False
+
+    def update_next_waypoint(self):
+        it = 0
+        nb_waypoints = len(self.static_waypoints.waypoints)
+        while not self.waypoint_is_ahead(self.static_waypoints.waypoints[self.next_waypoint % nb_waypoints]) and \
+                it < UPDATE_MAX_ITER:
+            it += 1
+            self.next_waypoint += 5  # We look at the next one
```
Comment (clockworkbits) [inline ros/src/waypoint_updater/waypoint_updater.py:101]: Should be `+1` ?

Comment (clockworkbits) [reply ros/src/waypoint_updater/waypoint_updater.py:101]: Extracted it to a separate issue https://github.com/clockworkbits/CarND-Capstone/issues/12

**Review by ghost — APPROVED** [review]
